File: src/libraries/CartesianClientLib/CartesianClient.cpp
// -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-

#include "CartesianClient.h"

/************************************************************************/
CartesianClient::CartesianClient() { }

/************************************************************************/
bool CartesianClient::open(const ConstString& serverPrefix) {
    ConstString clientStr(serverPrefix);
    clientStr += "/cartesianServer/rpc:o";
    rpcClient.open(clientStr);
    ConstString serverStr(serverPrefix);
    serverStr += "/cartesianServer/rpc:i";
    if (!rpcClient.addOutput(serverStr)) {
        fprintf(stderr,"[CartesianClient] error: no cartesianServer connection possible\n");
        return false;
    }
    printf("[CartesianClient] success: Opened connection with cartesianServer.\n");
    return true;
}

/************************************************************************/
bool CartesianClient::close() {
//    valid = false;
    printf("CartesianClient close...\n");
    rpcClient.interrupt();
    rpcClient.close();
    return true;
}

/************************************************************************/
bool CartesianClient::stat(double *x) {
    Bottle miOutput, miInput;
    miOutput.clear();
    miOutput.addVocab(VOCAB_STAT);
    rpcClient.write(miOutput, miInput);
//    printf("Got %s\n",miInput.toString().c_str());
    Bottle* data = miInput.get(1).asList();
    for (int i=0; i<data->size(); i++)
        x[i] = data->get(i).asDouble();
    return true;
}

/************************************************************************/
bool CartesianClient::inv(const double *xd, double *q) {
    Bottle miOutput, miInput;
    miOutput.clear();
    miOutput.addVocab(VOCAB_INV);
    Bottle dBottle;
    for(int i=0;i<NUM_AXES;i++)
        dBottle.addDouble(xd[i]);
    miOutput.addList() = dBottle;
    rpcClient.write(miOutput, miInput);
    Bottle* data = miInput.get(0).asList();
    for (int i=0; i<data->size(); i++)
        q[i] = data->get(i).asDouble();
    return true;
}

/************************************************************************/
bool CartesianClient::movj(const double *xd) {
    Bottle miOutput, miInput;
    miOutput.clear();
    miOutput.addVocab(VOCAB_MOVJ);
    Bottle dBottle;
    for(int i=0;i<NUM_AXES;i++)
        dBottle.addDouble(xd[i]);
    miOutput.addList() = dBottle;
    rpcClient.write(miOutput, miInput);
    return true;
}

/************************************************************************/
bool CartesianClient::movl(const double *xd) {
    Bottle miOutput, miInput;
    miOutput.clear();
    miOutput.addVocab(VOCAB_MOVL);
    Bottle dBottle;
    for(int i=0;i<NUM_AXES;i++)
        dBottle.addDouble(xd[i]);
    miOutput.addList() = dBottle;
    rpcClient.write(miOutput, miInput);
    return true;
}

/************************************************************************/
bool CartesianClient::wait() {
    Bottle miOutput, miInput;
    miOutput.clear();
    miOutput.addVocab(VOCAB_WAIT);
    rpcClient.write(miOutput, miInput);
    return true;
}

/************************************************************************/
bool CartesianClient::stop() {
    Bottle miOutput, miInput;
    miOutput.clear();
    miOutput.addVocab(VOCAB_MY_STOP);
    rpcClient.write(miOutput, miInput);
    return true;
}

/************************************************************************/

File: example/python/CartesianClient.py

import yarp

VOCAB_STAT = yarp.Vocab_encode('stat')
VOCAB_INV = yarp.Vocab_encode('inv')
VOCAB_MOVJ = yarp.Vocab_encode('movj')
VOCAB_MOVL = yarp.Vocab_encode('movl')
VOCAB_MY_STOP = yarp.Vocab_encode('stop')
VOCAB_WAIT = yarp.Vocab_encode('wait')

class CartesianClient:
    p = yarp.Port()
    def open(self, prefix):
        clientPort = prefix
        clientPort += '/cartesianClient/rpc:o'
        self.p.open(clientPort)
        serverPort = prefix
        serverPort += '/cartesianServer/rpc:i'
        self.p.addOutput(serverPort)
        return True
    def close(self):
        self.p.interrupt()
        self.p.close()
        return True
    def stat(self, res):
        miOutput = yarp.Bottle()
        miInput = yarp.Bottle()
        miOutput.clear()
        miOutput.addVocab(VOCAB_STAT)
        self.p.write(miOutput, miInput)
        data = miInput.get(0).asList()
        del res[:]
        for elem in range(0,data.size()):
            res.append(data.get(elem).asDouble())
        return True
    def stop(self):
        miOutput = yarp.Bottle()
        miInput = yarp.Bottle()
        miOutput.clear()
        miOutput.addVocab(VOCAB_MY_STOP)
        self.p.write(miOutput, miInput)
        return True
    def inv(self, xd, res):
        miOutput = yarp.Bottle()
        miInput = yarp.Bottle()
        miOutput.clear()
        miOutput.addVocab(VOCAB_INV)
        dBottle = yarp.Bottle()
        dBottle = miOutput.addList()
        for elem in range(0,len(xd)):
            dBottle.addDouble(xd[elem])
        self.p.write(miOutput, miInput)
        data = miInput.get(0).asList()
        del res[:]
        for elem in range(0,data.size()):
            res.append(data.get(elem).asDouble())
        return True
    def movj(self, xd):
        miOutput = yarp.Bottle()
        miInput = yarp.Bottle()
        miOutput.clear()
        miOutput.addVocab(VOCAB_MOVJ)
        dBottle = yarp.Bottle()
        dBottle = miOutput.addList()
        for elem in range(0,len(xd)):
            dBottle.addDouble(xd[elem])
        self.p.write(miOutput, miInput)
        return True
    def movl(self, xd):
        miOutput = yarp.Bottle()
        miInput = yarp.Bottle()
        miOutput.clear()
        miOutput.addVocab(VOCAB_MOVL)
        dBottle = yarp.Bottle()
        dBottle = miOutput.addList()
        for elem in range(0,len(xd)):
            dBottle.addDouble(xd[elem])
        self.p.write(miOutput, miInput)
        return True
    def wait(self):
        miOutput = yarp.Bottle()
        miInput = yarp.Bottle()
        miOutput.clear()
        miOutput.addVocab(VOCAB_WAIT)
        self.p.write(miOutput, miInput)
        return True

